### PR TITLE
Fix internet radio playback on bose soundtouch 10

### PR DESCRIPTION
**Pull request by GameTec-live**
## Description

I wanted to setup this project for my bose soundtouch 10 and while everything seemed to connect correctly, the internet radio playback, my main usecase, just didnt work.

This PR should fix playback (did so in my testing) and implements the needed endpoint(s).

Full transparency: This PR is 100% AI

## Type of Change

Please check the type of change your PR introduces:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ x ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Performance improvement
- [ ] Code refactoring (no functional changes)
- [ ] Test improvements
- [ ] Build/CI improvements

### API Changes
- [ x ] Added new endpoints
- [ x ] Modified existing endpoints
- [ ] Added new CLI commands
- [ ] Modified existing CLI commands
- [ ] Added new configuration options

### Implementation Details
- Describe the main changes
- List any new dependencies
- Mention any architectural changes

### Automated Tests
- [ ] Unit tests added/updated
- [ ] Integration tests added/updated
- [ ] All existing tests pass
- [ ] Test coverage maintained or improved

### Manual Testing
- [ ] Tested with real SoundTouch device(s)
- [ ] Tested CLI changes manually
- [ ] Tested in different network environments

**Device(s) tested with:**
- Device model: [e.g. SoundTouch 10]
- Device IP: [e.g. 192.168.1.100]
- Test results: [brief description]

### Test Commands
```bash
# Commands used to test this change
make test
go test ./pkg/client -v -run TestNewFeature
soundtouch-cli --host 192.168.1.100 new-command
```

## Documentation

- [ ] Updated relevant documentation
- [ ] Added code comments for complex logic
- [ ] Updated CLI help text
- [ ] Added usage examples
- [ ] Updated API documentation

**Documentation files updated:**
- [ ] README.md
- [ ] docs/API-Endpoints-Overview.md
- [ ] docs/CLI-REFERENCE.md
- [ ] Code documentation (godoc)

## Backward Compatibility

- [ ] This change is backward compatible
- [ ] This change includes breaking changes (requires major version bump)
- [ ] This change requires configuration migration

**Breaking changes (if any):**
- Describe what breaks
- Provide migration instructions

## Security Considerations

- [ ] No security implications
- [ ] Security review required
- [ ] Added input validation
- [ ] Updated authentication/authorization

## Performance Impact

- [ ] No performance impact
- [ ] Performance improvement
- [ ] Potential performance regression (justify why)

**Performance notes:**
- Measured impact: [benchmarks, timing, memory usage]
- Optimization opportunities: [if any]

## Code Quality

- [ ] Code follows project style guidelines
- [ ] No linting errors
- [ ] No security warnings
- [ ] Memory leaks checked (if applicable)

### Pre-submission Checklist

- [ ] `make check` passes (format, lint, vet)
- [ ] `make test` passes
- [ ] No TODO comments left in production code
- [ ] Error handling is comprehensive
- [ ] Logging is appropriate (not too verbose, not too quiet)

## Deployment Notes

Any special considerations for deployment:
- Configuration changes required
- Database migrations needed
- Service restart required
- Rollback procedures

## Screenshots (if applicable)

If this PR includes UI changes or CLI output changes, include screenshots or terminal output examples.

```bash
# Before
$ soundtouch-cli old-command
Old output...

# After  
$ soundtouch-cli new-command
New improved output...
```

## Additional Notes

Any additional information that reviewers should know:
- Design decisions and trade-offs
- Future work planned
- Alternative approaches considered
- References to external documentation

## Review Requests

**Areas that need special attention:**
- [ ] Error handling logic
- [ ] Performance critical sections
- [ ] Security implications
- [ ] API design choices
- [ ] Documentation clarity

**Specific questions for reviewers:**
1. Question about design choice X?
2. Is error handling sufficient in section Y?
3. Should we consider alternative approach Z?

---

**Reviewer Guidelines:**
- Check that all tests pass
- Verify documentation is updated
- Test manually if device access available  
- Consider backward compatibility
- Evaluate error handling and edge cases

### CODE REVIEW

**Comment by gesellix** [comment]
Can you provide an example request or cli command for the internet radio you wanted to stream?

**Comment by GameTec-live** [comment]
I have tried plenty both on firmware 21 (which was on my speakers before) and 27 (latest official from bose).

I tried the ones migrated / already on the device, I tried ö3, ö1, ..., I tried the one example from the preset docs. (soundtouch-cli --host 192.168.1.100 preset store --slot 3 --source TUNEIN --location "/v1/playback/station/s33828" --name "K-LOVE Radio")

Interestingly the server seems to respond correctly (opening the returned stream in a browser works), but the ST10 just stays silent and blinks amber (cloud error).

Somehow, choosing a different stream and proxying parts work tho.

If you are willing to work with me on a cleaner solution, please, I am willing to help where possible. (My grandparents needed the kitchen radio working again tho, so letting AI bang against the wall and write a fix was faster to just get it running lol)

**Comment by GameTec-live** [comment]
It seems like the real issue is that the root certificates on the ST10  are expired. It cant connect to any https site without throwing https errors.

As the server returns a https stream url, thats probably the issue and why a proxy resolves it.

Ig we either need to find a way to shove new certs into there or find a way to cleanly proxy the stream?

File: pkg/service/bmx/bmx.go [hunk 1]
```diff
@@ -660,6 +661,86 @@ func TuneInPlayback(stationID string) (*models.BmxPlaybackResponse, error) {
...
+
+func tuneInDirectStreamVariant(rawURL string) (string, bool) {
+	u, err := url.Parse(rawURL)
+	if err != nil {
+		return "", false
+	}
+
+	basePath := ""
+	switch {
+	case strings.HasSuffix(u.Path, "/hls"):
```
Comment (gesellix) [inline pkg/service/bmx/bmx.go:697]: Not sure if https://github.com/gesellix/Bose-SoundTouch/pull/249 is related to that.

**Comment by gesellix** [comment]
Please share more details about your AfterTouch version, logs from speaker and the soundtouch-service. Please ensure that your running a recent version of AfterTouch.

About your question regarding TLS/HTTPS issues: that's in fact highly probable and I'm going to verify that theory using the speaker's trust chain. If the issue is confirmed, I'm going to add a feature which allows you to add new/more recent trust certificates.

The other code changes are not that easy to merge, but let's talk about that when the TLS issues are addressed and maybe we don't need those code changes anymore.

**Comment by gesellix** [comment]
Do you have access to the ST10 via SSH?

**Comment by GameTec-live** [comment]
I do indeed have access to the st10 via ssh, how else would I install aftertouch lol

I pulled the latest release, v0.80.0

Sadly all I can tell you is that the requests get handled properly (200) as, without one exception (there I managed to confirm that the URLs are correct), the service refuses to log any bodies or responses even if explicitly enabled.

If you dont have a ST10, id still be happy to help tho, just lmk what details and logs from the speaker you need and where to find them.

I have since also hacked together a mini version that can run on the speaker itself and has just enough smarts for internet radio, so all I need, and proxying the stream clearly doesn't seem to be the right solution, so Ill just close this pr.

**Comment by gesellix** [comment]
Thanks. I'm going to make /hls to be opt-in. That seems to be the reason for your issues. Cert validation are no issue on the 27.x firmware.

**Comment by GameTec-live** [comment]
Cool, thanks.

Interesting tho, that any commands, wget, curl, etc that reach out to any https site fail the cert validation. Idk, maybe just my devices being weird? Date/Time not being synced correctly? Who knows.

**Comment by gesellix** [comment]
@GameTec-live I'm going to remove the hls stream format for the next release. See https://github.com/gesellix/Bose-SoundTouch/pull/300 for context. You might retry then based on the new version.

**Comment by gesellix** [comment]
> Cool, thanks.
> 
> Interesting tho, that any commands, wget, curl, etc that reach out to any https site fail the cert validation. Idk, maybe just my devices being weird? Date/Time not being synced correctly? Who knows.

Feel free to share a failing curl example, I'm happy to help!

**Comment by gesellix** [comment]
Examples using `curl` with the speaker's cert chain: https://github.com/gesellix/Bose-SoundTouch/pull/301

**Comment by GameTec-live** [comment]
Thanks a lot for the docs and fixes!

And thanks a lot for making this awesome project and keeping these devices running.

While I probably wont be using it now (save a rpi and my grandparents only need radio lol), this project was both base and inspiration for my hacked together on device version (https://github.com/GameTec-live/soundtouch-tiny).

**Comment by gesellix** [comment]
That's nice! I'm going to check your project to provide a similar "tiny" variant here. I hope you're fine with me taking inspiration back here ;)

**Comment by GameTec-live** [comment]
Oh please! Again, as this "only need to work once and for me", this is mostly AI, but the idea was, that theres still a few resources free on these embedded systems, so lets just let the speaker talk with itself.

Idk how many features one could actually cram in and I wanted it to be as simple as possible, so only internet radio is implemented. No config required, the thing just generates bogus values to make the speaker happy.

Theres a helper config command to help one configure the presets with the radio stations one likes and the migrate command sets everything up (ssh persistence, programm (start before boses software), optional config, change over servers in xml and turn the usb "service" port into a serial console to allow for easier config when wifi isnt available).

Id love to see you or someone else refine that, reimplement it properly and maybe add back stuff like spotify and whatever else one might need.

You might want to omit the serial console feature. While handy in my case, setting up the speakers when the final wifi isnt available and in general just having a fallback, the kernel would really prefer if it stays an ethernet gadget lol

For example; The VID and PID still say ethernet gadget, nomatter what I try, so windows refuses to load serial drivers. Under linux it works fine tho.

It might also be worth investigating if a full out firmware replacement might be possible. Its a standard TI SOC, but considering they still go for 100€ used here (like wtf, what normal person is buying speakers from 2015 with a dead cloud), Ill have to see what I can do.

If you want to maybe collab a bit more or go more into this "all on device" direction, feel free to hit me up on Discord (my server or dms, both fine), email or via paper receipt.
